compiler/rustc_hir_analysis/src/check/compare_method.rs

```diff
@@ -342,6 +346,31 @@ fn compare_predicate_entailment<'tcx>(
     Ok(())
 }
 
+fn compare_asyncness<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl_m: &AssocItem,
+    impl_m_span: Span,
+    trait_m: &AssocItem,
+    trait_item_span: Option<Span>,
+) -> Result<(), ErrorGuaranteed> {
+    if tcx.asyncness(trait_m.def_id) == hir::IsAsync::Async {
+        if tcx.asyncness(impl_m.def_id) != hir::IsAsync::Async {
+            let mut multispan = MultiSpan::from(impl_m_span);
+            if let Some(trait_item_span) = trait_item_span {
+                multispan
+                    .push_span_label(trait_item_span, "required because the trait method is async");
+            }
+
+            let mut diag = tcx
+                .sess
+                .struct_span_err(multispan, &format!("method `{}` should be async", trait_m.name));
```

Comment:
Do you mind making this into a struct with `derive(Diagnostic)`? https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-structs.html

Reply:
The primary message of the error should explain itself fully without more context, so "because the method it's implementing from the trait is `async`" or similar should probably be appended here.

It's fine for the label to duplicate that "because" part, that's common for errors.